libsupport/include/katana/HostAllocator.h

```diff
@@ -0,0 +1,37 @@
+#pragma once
+
+#include <cstdlib>
+#include <cstring>
+
+#include "katana/config.h"
+
+namespace katana {
+
+enum class MemoryPinType { Swappable = 0, Pinned = 1 };
+
+class BaseHostAllocator {
+public:
+  virtual void* Malloc(const size_t n_bytes) const = 0;
+  virtual void* Calloc(const size_t n_items, const size_t item_size) const = 0;
+  virtual void* Realloc(void* ptr, const size_t n_bytes) const = 0;
+  virtual void Free(void* ptr) const = 0;
+  virtual MemoryPinType GetType() const = 0;
+  virtual ~BaseHostAllocator() {}
+};
+
+class SwappableHostAllocator : public BaseHostAllocator {
+public:
+  void* Malloc(const size_t n_bytes) const override { return malloc(n_bytes); }
+  void* Calloc(const size_t n_items, const size_t item_size) const override {
+    return calloc(n_items, item_size);
+  }
+  void* Realloc(void* ptr, const size_t n_bytes) const override {
+    return realloc(ptr, n_bytes);
+  }
+  void Free(void* ptr) const override { free(ptr); }
+  MemoryPinType GetType() const override { return MemoryPinType::Swappable; }
+};
+
+KATANA_EXPORT extern const SwappableHostAllocator swappable_host_allocator;
```

Comment:
Should this be a method which return `const HostAllocator*` or `const HostAllocator&`? That way we can change the actual implementation of this default dynamically in the future if we need to. It would help a lot for debugging (using a special tracing allocator), and for integrating with other systems with allocators that we need to use everywhere.

Reply:
Yes, it should be a function implemented in .cpp file that returns a non-const pointer or reference.